cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java

```diff
@@ -83,8 +91,12 @@ private RouteDestination getV3RoutingService(String [] uriParts, AllowedMethod r
       //       <user-defined-method-path>"
       String serviceName = ServiceDiscoverable.getName(uriParts[2], uriParts[4], uriParts[8]);
       String version = uriParts[6];
-      // TODO: Optimize this instead of creating new object every time
-      return new RouteDestination(serviceName, version);
+      MultiKey cacheKey = new MultiKey(serviceName, version);
+
+      if (!routeDestinationCache.containsKey(cacheKey)) {
```

Comment:
hmmm, why do you need a synchronized map? Also, if you do a ``containsKey`` + ``put``, its not really synchronized. Use ``putIfAbsent``.